deepspeed/pt/deepspeed_zero_optimizer.py

```diff
@@ -1087,10 +1199,160 @@ def free_grad_in_param_list(self, param_list):
+        start = self.first_offset[group_index]
+        end = start + self.partition_size[group_index]
+        return [int(start), int(end)]
+
+    #jie:
+    def step_with_cpuoffload(self, closure=None):
+        """
+        Not supporting closure.
+        """
+        see_memory_usage(f"In step before checking overflow")
+
+        # First compute norm for all group so we know if there is overflow
+        self.check_overflow()
+
+        timers = self.timers
+
+        prev_scale = self.loss_scale
+        self._update_scale(self.overflow)
+        if self.overflow:
+            see_memory_usage('After overflow before clearing gradients')
+            self.zero_grad()
+            see_memory_usage('After overflow after clearing gradients')
+
+            logger.info(
+                "[deepscale] OVERFLOW! Rank {} Skipping step. Attempted loss scale: {}, "
+                "reducing to {}".format(dist.get_rank(),
+                                        prev_scale,
+                                        self.loss_scale))
+            timers('optimizer_step').start()
+            timers('optimizer_step').stop()
+            timers('optimizer_allgather').start()
+            timers('optimizer_allgather').stop()
+            return
+
+        norm_groups = []
+        single_partition_grad_groups = []
+        skip = False
+        partition_id = dist.get_rank(group=self.dp_process_group)
+        for i, group in enumerate(self.fp16_groups):
+
+            norm_groups.append(
+                self.get_grad_norm_direct(self.averaged_gradients[i],
+                                          self.params_in_partition[i]))
+
+            #free gradients for all the prameters that are not updated by this process
+            self.free_grad_in_param_list(self.params_not_in_partition[i])
+
+            #create a flat gradients for parameters updated by this process
+            # If we are last partition, ensure we have same size grads and partition size, if not pad with zero tensors
+            if partition_id == dist.get_world_size(group=self.dp_process_group) - 1:
+                single_grad_partition = flatten_dense_tensors_aligned(
+                    self.averaged_gradients[i],
+                    int(self.partition_size[i])).to(
+                        self.single_partition_of_fp32_groups[i].dtype)
+            else:
+                single_grad_partition = _flatten_dense_tensors(
+                    self.averaged_gradients[i]).to(
+                        self.single_partition_of_fp32_groups[i].dtype)
+            assert single_grad_partition.numel() == self.partition_size[i], \
+                "averaged gradients have different number of elements that partition size {} {} {} {}".format(single_grad_partition.numel(), self.partition_size[i], i, partition_id)
+
+            #jie: transform the grad back to CPU
+            start, end = self.get_partition_fp32_group_index(i)
+            with torch.cuda.stream(self.migration_stream):

```

Comment:
I doubt this is overlapping with anything. This needs to be done directly during the backward pass if we want to overlap it with computation. We may want to revisit this to understand the performance implication. 

It probably ok for this PR though. Please just mark it as a TODO